src/instructlab/sdg/sdg.py

```diff
@@ -1,21 +1,88 @@
 # SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Optional
+
 # Third Party
-from datasets import Dataset
+from datasets import Dataset, concatenate_datasets
 
 # Local
 from .pipeline import Pipeline
 
 
 # This is part of the public API.
 class SDG:
-    def __init__(self, pipelines: list[Pipeline]) -> None:
+    def __init__(
+        self,
+        pipelines: list[Pipeline],
+        num_workers: Optional[int] = 1,
+        batch_size: Optional[int] = None,
+    ) -> None:
+        """
+        Initialize a new instance of SDG with a collection of pipelines to run
+        in sequence.
+
+        pipelines (list[Pipeline]): A list of pipeline objects to run in
+            sequence.
+        num_workers (int | None): The number of workers to use. None implies the
+            default behavior of min(32, os.cpu_count() + 4)
+            https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+        batch_size (int | None): The batch size to use when processing. None
+            implies no batching.
+        """
         self.pipelines = pipelines
+        self.num_workers = num_workers
+        self.batch_size = batch_size
 
     def generate(self, dataset: Dataset):
         """
         Generate the dataset by running the chained pipeline steps.
-        dataset: the input dataset
+        dataset (Dataset): the input dataset
+        """
+        input_splits = self._split_dataset(dataset)
+        output_splits = [None] * len(input_splits)
+
+        with ThreadPoolExecutor(max_workers=self.num_workers) as executor:
+            futures = [
+                executor.submit(
+                    self._generate_split, self.pipelines, input_split, output_splits, i
```

Comment:
I'd move all of this concurrency code into `Pipeline` itself - I don't the benefit of it being a separate abstraction

Reply:
Interesting. The only benefit I see for keeping it in SDG is that the sharding happens once and then each shard flows the entire thing independently. If this moves to `Pipeline`, it will essentially force a sync between chained `Pipeline`s. I think I remember there being some discussion elsewhere about whether the `SDG` class was needed since it's essentially a pipeline of pipelines. Was there a conclusion to that discussion? I could certainly see fully smashing `SDG` into `Pipeline` and avoiding this wrapping, but if `SDG` remains, I think there's a _slight_ advantage to having the batching live here.

Reply:
see https://github.com/instructlab/dev-docs/pull/113

we are likely to add a higher level "router" concept, unrelated - https://github.com/instructlab/sdg/issues/134

I'll be resisting extending SDG functionality without renaming it to a  descriptive noun